### PR TITLE
Cache latest versions in Redis

### PR DESCRIPTION
**Pull request by abouthiroppy**
Hello :)

## issue
  https://github.com/typings/api/issues/35

## added files
  - `src/api/routes/support/redis.ts`
  - `src/worker/jobs/support/redis.ts`

## added libraries
  - `lodash`
  - `redis`

## redisKey
I define redisKey is `{source}:{name}`

## schema (`version/latest`)
```
{
  "tag": "1.5.0+20160318102130",
  "version": "1.5.0",
  "description": null,
  "compiler": null,
  "location":     "github:DefinitelyTyped/DefinitelyTyped/angularjs/angular.d.ts#36f1d8d82f0593d839aad6944260fd19a090aa78",
  "updated": "2016-03-18T10:21:30.000Z",
  "deprecated": null
}
```

```
{
  "tag": "1.5.0+20160318102130",
  "version": "1.5.0",
  "location":   "github:DefinitelyTyped/DefinitelyTyped/angularjs/angular.d.ts#36f1d8d82f0593d839aad6944260fd19a090aa78",
  "updated": "2016-03-18T10:21:30.000Z",
  "deprecated": null
}
```

in redis
```
...
2398) "dt:shuffle-array"
2399) "dt:definitions/unitywebapi"
2400) "npm:pretty-data"
2401) "dt:js-data/legacy/js-data"
2402) "dt:notify.js"
2403) "dt:jquery.nouislider"
2404) "dt:envify"
2405) "dt:d3"
2406) "dt:soap"
```
@blakeembrey  In the case of `undefined`, a property disappears in the schema. This is because my Implementation uses `JSON.stringify`. When the property disappears, is the influence given?

I think that we should create the environment to write a test.

thanks.

### CODE REVIEW

File: src/api/routes/support/db.ts [hunk 1]
```diff
@@ -121,6 +122,16 @@ export function getMatchingVersions (source: string, name: string, version: stri
  * Retrieve the latest available version.
  */
 export function getLatest (source: string, name: string, version?: string) {
+  return getVersionInRedis(source, name).then((data) => {
```
Comment (blakeembrey) [inline src/api/routes/support/db.ts]: This doesn't work with the version operator. If it's in Redis, searching by version will break.

File: src/worker/jobs/support/db.ts [hunk 2]
```diff
@@ -12,9 +16,17 @@ export interface UpsertOptions {
   trx?: knex.Transaction
   where?: string
   returning?: string[]
+  redisKey?: string
 }
 
 export function upsert (options: UpsertOptions): Promise<Object> {
+  if (options.table === 'versions') {
```
Comment (blakeembrey) [inline src/worker/jobs/support/db.ts]: I wouldn't so it this way. I'd only have jobs mark things as invalid (E.g. delete entries), but I'd make the front-end APIs handing caching. This way the entire project doesn't need to be re-indexed to update the cache, we just mark things as invalid (or delete them, in this case).

**Comment by blakeembrey** [comment]
I made some comments, but I'd invert how control occurs here. Have the API check Redis first, then fallback to DB and persist the result. Only have jobs invalidating entries. This allows deployments and re-indexing to occur independent of the cache layer. Otherwise any changes in the schema need to re-index all jobs.

**Comment by abouthiroppy** [comment]
Thank you for posting comments :)
Because it seems to be difficult, I close this PR.

**Pull request closed without merging**